File: JavaScript/2407/240702/BAEK_2503.js
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

const recur = (hintIdx, number) => {
  if (hintIdx === N) {
    answer += 1;
    recur(0, number+1);
    return;
  }
  if (number > 999) return;

  if (checker(hintIdx, number)) {
    recur(hintIdx+1, number);
  } else {
    recur(0, number+1);
  }
}

const checker = (hintIdx, number) => {
  strNum = String(number);

  if (strNum.includes('0')) {
    return false;
  }

  const numSet = new Set();
  for (const char of strNum) {
    if (numSet.has(char)) {
      return false;
    } else {
      numSet.add(char);
    }
  }

  let strikeCount = 0;
  let ballCount = 0;

  const strHint = String(hints[hintIdx][0]);

  if (strNum[0] === strHint[0]) {
    strikeCount += 1;
  } else if (strNum[0] === strHint[1] || strNum[0] === strHint[2]) {
    ballCount += 1;
  }
  if (strNum[1] === strHint[1]) {
    strikeCount += 1;
  } else if (strNum[1] === strHint[0] || strNum[1] === strHint[2]) {
    ballCount += 1;
  }
  if (strNum[2] === strHint[2]) {
    strikeCount += 1;
  } else if (strNum[2] === strHint[0] || strNum[2] === strHint[1]) {
    ballCount += 1;
  }

  if (strikeCount === hints[hintIdx][1] && ballCount === hints[hintIdx][2]) {
    return true;
  } else {
    return false;
  }
}

let N;
const hints = [];
let idx = 0;
let answer = 0;

rl.on('line', (line) => {
  if (idx === 0) {
    N = parseInt(line);
  } else {
    hints.push(line.split(' ').map(Number));
  }
  idx++;
  if (idx > N) {
    rl.close();
  }
}).on('close', () => {
  recur(0, 100);
  console.log(answer);
});

File: Python/2024/2407/240702/BAEK_2503.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline


def recur(hint_idx, number):
    global answer
    if hint_idx == N:
        answer += 1
        recur(0, number+1)
        return

    # 1000 미만
    if number > 999:
        return

    # 중복 제거
    str_num = str(number)
    if str_num[0] == '0' or str_num[1] == '0' or str_num[2] == '0':
        recur(0, number+1)
        return
    if str_num[0] == str_num[1] or str_num[0] == str_num[2] or str_num[1] == str_num[2]:
        recur(0, number+1)
        return

    # 힌트 검사
    str_hint = str(hints[hint_idx][0])
    strike_cnt = 0
    ball_cnt = 0

    if str_num[0] == str_hint[0]:
        strike_cnt += 1
    elif str_num[0] == str_hint[1] or str_num[0] == str_hint[2]:
        ball_cnt += 1
    if str_num[1] == str_hint[1]:
        strike_cnt += 1
    elif str_num[1] == str_hint[0] or str_num[1] == str_hint[2]:
        ball_cnt += 1
    if str_num[2] == str_hint[2]:
        strike_cnt += 1
    elif str_num[2] == str_hint[0] or str_num[2] == str_hint[1]:
        ball_cnt += 1

    if strike_cnt == hints[hint_idx][1] and ball_cnt == hints[hint_idx][2]:
        # 힌트 통과
        recur(hint_idx+1, number)
    else:
        # 힌트 미통과
        recur(0, number+1)


N = int(input())
hints = [list(map(int, input().split())) for _ in range(N)]
answer = 0
recur(0, 100)
print(answer)
